bundles/org.openhab.binding.androidtv/src/main/java/org/openhab/binding/androidtv/internal/protocol/shieldtv/ShieldTVConnectionManager.java

```diff
@@ -0,0 +1,1218 @@
+        if (!isLoggedIn) {
+            isOnline = servicePing();
+        } else {
+            isOnline = true;
+        }
+        logger.debug("{} - Device Health - Online: {} - Logged In: {}", handler.getThingID(), isOnline, isLoggedIn);
+        if (isOnline != this.isOnline) {
+            this.isOnline = isOnline;
+            if (isOnline) {
+                logger.debug("{} - Device is back online.  Attempting reconnection.", handler.getThingID());
+                reconnect();
+            }
+        }
+    }
+
+    public void setKeys(String privKey, String cert) {
+        try {
+            androidtvPKI.setKeys(privKey, encryptionKey, cert);
+            androidtvPKI.saveKeyStore(config.keystorePassword, encryptionKey);
+        } catch (GeneralSecurityException e) {
+            logger.debug("General security exception", e);
+        } catch (IOException e) {
+            logger.debug("IO Exception", e);
+        } catch (Exception e) {
+            logger.debug("General Exception", e);
+        }
+    }
+
+    public void setAppDB(Map<String, String> appNameDB, Map<String, String> appURLDB) {
+        this.appNameDB = appNameDB;
+        this.appURLDB = appURLDB;
+        this.appDBPopulated = true;
+        logger.debug("{} - App DB Populated", handler.getThingID());
+        logger.trace("{} - Handler appNameDB: {} appURLDB: {}", handler.getThingID(), this.appNameDB.toString(),
+                this.appURLDB.toString());
+        handler.updateCDP(CHANNEL_APP, this.appNameDB);
+    }
+
+    private TrustManager[] defineNoOpTrustManager() {
+        return new TrustManager[] { new X509TrustManager() {
+            @Override
+            public void checkClientTrusted(final X509Certificate @Nullable [] chain, final @Nullable String authType) {
+                logger.debug("Assuming client certificate is valid");
+                if (chain != null && logger.isTraceEnabled()) {
+                    for (int cert = 0; cert < chain.length; cert++) {
+                        logger.trace("Subject DN: {}", chain[cert].getSubjectX500Principal());
+                        logger.trace("Issuer DN: {}", chain[cert].getIssuerX500Principal());
+                        logger.trace("Serial number {}:", chain[cert].getSerialNumber());
+                    }
+                }
+            }
+
+            @Override
+            public void checkServerTrusted(final X509Certificate @Nullable [] chain, final @Nullable String authType) {
+                logger.debug("Assuming server certificate is valid");
+                if (chain != null && logger.isTraceEnabled()) {
+                    for (int cert = 0; cert < chain.length; cert++) {
+                        logger.trace("Subject DN: {}", chain[cert].getSubjectX500Principal());
+                        logger.trace("Issuer DN: {}", chain[cert].getIssuerX500Principal());
+                        logger.trace("Serial number: {}", chain[cert].getSerialNumber());
+                    }
```

Comment:
Can it be extracted to a method and reused instead of being duplicated?

Reply:
To be honest, I've had a ton of issues with the TrustManagers across this whole binding.  They are very finicky and the only thing that actually matters is the getAcceptedIssuers() function.  The two check functions are just for debugging.  Can I add this as a to-do for later as I go through and make the code more efficient?